commit c7afbdc3ea5290f10de43e0d9652de5bac27539b
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-07-30T21:23:20+00:00 (10 weeks ago)
    
    MANTA-4440 manta-deploy-dev no longer idempotent
    Reviewed by: Trent Mick <trentm@gmail.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/bin/manta-deploy-dev b/bin/manta-deploy-dev
index 4f412e1..9237639 100755
--- a/bin/manta-deploy-dev
+++ b/bin/manta-deploy-dev
@@ -41,6 +41,9 @@ function confirm
 	[[ $REPLY =~ ^[Yy]$ ]]
 }
 
+# Must be kept in sync with the value in lib/common.js
+RING_EXISTS_EXIT_STATUS=3
+
 mdl_tmpdir="${TMPDIR:-/var/tmp}"
 mdl_conffile="$mdl_tmpdir/lab-config.json"
 mdl_override="false"		# override confirmation prompts
@@ -85,6 +88,9 @@ manta-shardadm set -s "1.moray"
 set +o errexit
 manta-adm create-topology -v $mdl_topo_size -p 2020
 case "$?" in
+	${RING_EXISTS_EXIT_STATUS})
+		echo "hash ring already exists"
+		;;
 	0)
 		echo "successfully generated hash ring"
 		;;
diff --git a/cmd/manta-adm.js b/cmd/manta-adm.js
index bfb4965..98a9db3 100755
--- a/cmd/manta-adm.js
+++ b/cmd/manta-adm.js
@@ -2368,10 +2368,8 @@ MantaAdm.prototype.do_create_topology = function (subcmd, opts, args, callback)
 			port: opts.p,
 			force: force
 		}, function createdTopology(err) {
-			if (err) {
-				fatal(err.message);
-			}
 			self.finiAdm();
+			callback(err);
 		});
 	});
 };
diff --git a/lib/adm.js b/lib/adm.js
index 1e7d089..567ae56 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -4932,8 +4932,10 @@ maAdm.prototype.createTopology = function createTopology(opts, callback)
 		var key = buckets ? common.BUCKETS_HASH_RING_IMAGE :
 		    common.HASH_RING_IMAGE;
 		if (mantaApp.metadata[key] && !force) {
-			next(new VError(
-			    'Hash ring image already exists in SAPI metadata'));
+			var err = new VError(
+			    'Hash ring image already exists in SAPI metadata');
+			err.exitStatus = common.RING_EXISTS_EXIT_STATUS;
+			next(err);
 			return;
 		}
 		next();
diff --git a/lib/common.js b/lib/common.js
index 2f1094f..a2659ea 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -47,6 +47,13 @@ exports.HASH_RING_IMAGE = 'HASH_RING_IMAGE';
 exports.BUCKETS_HASH_RING_IMAGE = 'BUCKETS_HASH_RING_IMAGE';
 exports.HASH_RING_IMGAPI_SERVICE = 'HASH_RING_IMGAPI_SERVICE';
 
+/*
+ * A special process exit code to be returned by manta-adm `create-topology`
+ * if the hash ring already exists in SAPI. This is used by
+ * bin/manta-deploy-dev and must be kept in sync with the value there.
+ */
+exports.RING_EXISTS_EXIT_STATUS = 3;
+
 exports.CONFIG_FILE_DEFAULT = path.join(__dirname, '..', 'etc', 'config.json');
 
 exports.GC_METADATA_FIELDS = [
